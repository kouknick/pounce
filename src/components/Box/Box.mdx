Responsive box-model layout component.

```typescript jsx
<Box p={9} fontSize={4} width={300} color="white" bg="primary200">
  <Text size="medium" color="white" fontSize={4}>
    Box
  </Text>
</Box>
```
