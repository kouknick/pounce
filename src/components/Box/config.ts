import { createShouldForwardProp, props } from '@styled-system/should-forward-prop';

export const extraConfig = {
  direction: {
    property: 'flexDirection' as const,
  },
  justify: {
    property: 'justifyContent' as const,
  },
  align: {
    property: 'alignItems' as const,
  },
  shadow: {
    property: 'boxShadow' as const,
    scale: 'shadows',
  },
  fill: {
    property: 'fill' as const,
    scale: 'colors',
  },
  stroke: {
    property: 'stroke' as const,
    scale: 'colors',
  },
  viewBox: true,
  textDecoration: true,
  overflowX: true,
  overflowY: true,
  textTransform: true,
  animation: true,
  transform: true,
  visibility: true,
  whiteSpace: true,
  pointerEvents: true,
  wordBreak: true,
  overflowWrap: true,
  textOverflow: true,
  cursor: true,
  resize: true,
  transition: true,
  objectFit: true,
  outline: true,
  willChange: true,
};

// extend the forwarded props by stuff that styled-system doesn't deal with
export const shouldForwardProp = createShouldForwardProp([
  ...props,
  'textDecoration',
  'pointerEvents',
  'visibility',
  'transform',
  'cursor',
]);
