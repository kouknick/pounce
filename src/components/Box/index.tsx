export { default } from './Box';
export type { NativeAttributes, BoxProps } from './Box';
export * from './system';
