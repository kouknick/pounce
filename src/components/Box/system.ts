import React from 'react';
import { createShouldForwardProp, props } from '@styled-system/should-forward-prop';
import * as StyledSystem from 'styled-system';
import { Theme } from '../../theme';

// prettier-ignore
// Gather all the styled-system props that we are going to pass
type ThemedStyleProps =
  StyledSystem.SpaceProps<Theme, React.CSSProperties['margin']> &
  StyledSystem.ColorProps<Theme> &
  StyledSystem.LayoutProps<Theme> &
  StyledSystem.BackgroundProps<Theme> &
  StyledSystem.GridProps<Theme> &
  StyledSystem.ShadowProps<Theme> &
  StyledSystem.BorderProps<Theme> &
  StyledSystem.PositionProps<Theme> &
  StyledSystem.FlexboxProps<Theme> &
  StyledSystem.TypographyProps<Theme>;

// Props related to Link components in @reach/router & react-router
type RoutingProps = {
  /** The url path to navigate, if the component implements a Link
   * @ignore
   */
  to?: string;
};

// Props related to the usage of the Emotion CSS-in-JS library
type EmotionProps = {
  /** The React Component or native HTML element to render instead.
   * @default "div"
   * @ignore
   */
  as?: React.ElementType;
  /** Additional custom inline CSS to pass to the element
   * @ignore
   */
  css?: any;
};

// Gather the custom-named props that styled-system should accept
export interface CustomStyleProps {
  shadow?: ThemedStyleProps['boxShadow'];
  fill?: ThemedStyleProps['color'];
  stroke?: ThemedStyleProps['color'];
  textDecoration?: StyledSystem.ResponsiveValue<React.CSSProperties['textDecoration']>;
  textTransform?: StyledSystem.ResponsiveValue<React.CSSProperties['textTransform']>;

  animation?: StyledSystem.ResponsiveValue<React.CSSProperties['animation']>;
  transform?: StyledSystem.ResponsiveValue<React.CSSProperties['transform']>;
  visibility?: StyledSystem.ResponsiveValue<React.CSSProperties['visibility']>;
  whiteSpace?: StyledSystem.ResponsiveValue<React.CSSProperties['whiteSpace']>;
  pointerEvents?: StyledSystem.ResponsiveValue<React.CSSProperties['pointerEvents']>;
  wordBreak?: StyledSystem.ResponsiveValue<React.CSSProperties['wordBreak']>;
  overflowWrap?: StyledSystem.ResponsiveValue<React.CSSProperties['overflowWrap']>;
  textOverflow?: StyledSystem.ResponsiveValue<React.CSSProperties['textOverflow']>;
  cursor?: StyledSystem.ResponsiveValue<React.CSSProperties['cursor']>;
  resize?: StyledSystem.ResponsiveValue<React.CSSProperties['resize']>;
  transition?: StyledSystem.ResponsiveValue<React.CSSProperties['transition']>;
  objectFit?: StyledSystem.ResponsiveValue<React.CSSProperties['objectFit']>;
  outline?: StyledSystem.ResponsiveValue<React.CSSProperties['outline']>;
  willChange?: StyledSystem.ResponsiveValue<React.CSSProperties['willChange']>;
  borderCollapse?: StyledSystem.ResponsiveValue<React.CSSProperties['borderCollapse']>;
  tableLayout?: StyledSystem.ResponsiveValue<React.CSSProperties['tableLayout']>;
}

export const customStyleProps: Record<
  keyof CustomStyleProps,
  boolean | StyledSystem.ConfigStyle
> = {
  shadow: {
    property: 'boxShadow' as const,
    scale: 'shadows',
  },
  fill: {
    property: 'fill' as const,
    scale: 'colors',
  },
  stroke: {
    property: 'stroke' as const,
    scale: 'colors',
  },
  textDecoration: true,
  textTransform: true,
  animation: true,
  transform: true,
  visibility: true,
  whiteSpace: true,
  pointerEvents: true,
  wordBreak: true,
  overflowWrap: true,
  textOverflow: true,
  cursor: true,
  resize: true,
  transition: true,
  objectFit: true,
  outline: true,
  willChange: true,
  borderCollapse: true,
  tableLayout: true,
};

// All of the allowed props gathered together
export type SystemProps = ThemedStyleProps & CustomStyleProps & RoutingProps & EmotionProps;

// extend the forwarded props by stuff that styled-system doesn't deal with
export const shouldForwardProp = createShouldForwardProp([
  ...props,
  'textDecoration',
  'pointerEvents',
  'visibility',
  'transform',
  'cursor',
]);

// Transform the custom alias to a format that styled-system CSS supports
function transformCustomStyleAlias(
  propName: keyof CustomStyleProps,
  propValue: CustomStyleProps[keyof CustomStyleProps]
) {
  const customStylePropValue = customStyleProps[propName];
  if (typeof customStylePropValue === 'object') {
    return { [customStylePropValue.property as string]: propValue };
  }
  if (customStylePropValue === true) {
    return {
      [propName]: propValue,
    };
  }
  return {};
}

// Transform the custom alias to a format that styled-system CSS supports
function transformThemedAlias(
  propName: keyof ThemedStyleProps,
  propValue: ThemedStyleProps[keyof ThemedStyleProps]
) {
  return {
    [propName]: propValue,
  };
}

// Transform the custom alias to a format that styled-system CSS supports
const transformAlias = (
  propName: keyof ThemedStyleProps & keyof CustomStyleProps,
  propValue: ThemedStyleProps[keyof ThemedStyleProps] & CustomStyleProps[keyof CustomStyleProps]
) => {
  if (Object.keys(customStyleProps).includes(propName)) {
    return transformCustomStyleAlias(propName, propValue);
  }
  return transformThemedAlias(propName, propValue);
};

export const transformAliasProps = (props?: SystemProps): SystemProps => {
  let result = {};
  if (!props) {
    return result;
  }

  for (const propName in props) {
    // @ts-ignore
    const propValue = props[propName];
    if (typeof propValue === 'object' && !Array.isArray(propValue)) {
      result = { ...result, [propName]: transformAliasProps(propValue) };
    } else {
      result = {
        ...result,
        ...transformAlias(propName as keyof (ThemedStyleProps | CustomStyleProps), propValue),
      };
    }
  }
  return result;
};
