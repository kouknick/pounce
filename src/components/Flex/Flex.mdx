```typescript jsx
<Flex alignItems="center">
  <Box width={200} py={9} bg="blue-400">
    <Text size="medium" color="white" textAlign="center">
      We are
    </Text>
  </Box>
  <Box width={300} py={4} bg="red-300">
    <Text size="medium" color="white" textAlign="center">
      vertically aligned
    </Text>
  </Box>
</Flex>
```
