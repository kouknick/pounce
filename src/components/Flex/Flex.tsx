import React from 'react';
import Box, { BoxProps } from '../Box';

export type FlexProps = Omit<BoxProps, 'display'> & {
  /** An alias for `flexDirection` */
  direction?: BoxProps['flexDirection'];

  /** An alias for `justifyContent` */
  justify?: BoxProps['justifyContent'];

  /** An alias for `alignItems` */
  align?: BoxProps['alignItems'];

  /** Whether the flex should be `flex-inline` */
  inline?: boolean;
};

/**
 * Extends <a href="/#/Box">Box</a>
 *
 * Responsive flexbox layout component. You should use this anytime you want a flex container or
 * wrapper around a certain layout
 */
const Flex: React.FC<FlexProps> = React.forwardRef(function Flex(props, ref) {
  const { direction, justify, align, inline, ...rest } = props;
  return (
    <Box
      display={inline ? 'inline-flex' : 'flex'}
      ref={ref}
      flexDirection={direction}
      justifyContent={justify}
      alignItems={align}
      {...rest}
    />
  );
});

export default Flex;
