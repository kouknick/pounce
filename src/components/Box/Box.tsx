import React from 'react';
import styled from '../../system/styled';
import * as StyledSystem from 'styled-system';
import { customStyleProps, shouldForwardProp, SystemProps } from './system';

export type NativeAttributes<El extends React.ElementType> = Omit<
  React.ComponentPropsWithRef<El>,
  keyof SystemProps
>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BoxProps<Element extends React.ElementType = any> = NativeAttributes<Element> &
  SystemProps & {
    /** Whether should text should truncate to fill at most one line of text */
    truncated?: boolean;
  };

/**
 * Responsive box-model layout component. Apart from the defined props,
 * it also supports all the native HTML attributes.
 * */
const Box = styled('div');

export default Box as React.FC<BoxProps>;
