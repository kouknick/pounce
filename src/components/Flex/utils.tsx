import React from 'react';
import { FlexProps } from './Flex';

export const useItemSpacingProps = (
  spacing: FlexProps['margin'],
  direction: FlexProps['direction'],
  sx: FlexProps['sx']
): Partial<FlexProps> => {
  return React.useMemo(() => {
    if (!spacing) {
      return sx;
    }

    const isFlexColumn = direction && (direction as string).includes('column');
    return {
      sx: {
        ...sx,
        '& > *:not(:last-child)': {
          marginRight: isFlexColumn ? undefined : spacing,
          marginBottom: isFlexColumn ? spacing : undefined,
        },
      },
    };
  }, [spacing, direction, sx]);
};
