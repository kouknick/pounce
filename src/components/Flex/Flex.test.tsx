import React from 'react';
import { render } from 'test-utils';
import Flex from './Flex';
import Box from '../Box';

describe('Flex', () => {
  it('renders', async () => {
    const { container } = await render(<Flex />);
    expect(container).toMatchSnapshot();
  });

  it('works along with Box', async () => {
    const { container } = await render(
      <Flex alignItems="center">
        <Box width={200} py={9} bg="blue-400">
          Test
        </Box>
        <Box width={300} py={4} bg="red-300">
          Test
        </Box>
      </Flex>
    );
    expect(container).toMatchSnapshot();
  });
});
