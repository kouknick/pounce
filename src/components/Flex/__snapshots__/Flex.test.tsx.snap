// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex can specify both \`sx\` and custom spacing 1`] = `
.pounce-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.pounce-0 input {
  max-width: 5px;
}

.pounce-0>*:not(:last-child) {
  margin-right: 4px;
}

<div>
  <div
    class="pounce-0"
  />
</div>
`;

exports[`Flex renders 1`] = `
.pounce-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="pounce-0"
  />
</div>
`;

exports[`Flex works along with Box 1`] = `
.pounce-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.pounce-0 {
  width: 200px;
  padding-top: 9px;
  padding-bottom: 9px;
  background-color: blue-400;
}

.pounce-1 {
  width: 300px;
  padding-top: 32px;
  padding-bottom: 32px;
  background-color: red-300;
}

<div>
  <div
    class="pounce-2"
  >
    <div
      class="pounce-0"
    >
      Test
    </div>
    <div
      class="pounce-1"
    >
      Test
    </div>
  </div>
</div>
`;
